crates/sui-rest-api/src/transactions/resolve/literal.rs

```diff
@@ -0,0 +1,787 @@
+}
+
+fn determine_literal_type(
+    called_packages: &HashMap<ObjectId, NormalizedPackage>,
+    commands: &[Command],
+    arg_idx: usize,
+) -> Result<Type> {
+    fn set_type(maybe_type: &mut Option<Type>, ty: Type) -> Result<()> {
+        match maybe_type {
+            Some(literal_type) if literal_type == &ty => {}
+            Some(_) => {
+                return Err(RestError::new(
+                    axum::http::StatusCode::BAD_REQUEST,
+                    "unable to resolve literal as it is used as multiple different types across commands",
+                ))
+            }
+            None => {
+                *maybe_type = Some(ty);
+            }
+        }
+
+        Ok(())
+    }
+    let mut literal_type = None;
+
+    for (command, idx) in super::find_arg_uses(arg_idx, commands) {
+        match (command, idx) {
+            (Command::MoveCall(move_call), Some(idx)) => {
+                let arg_type = super::arg_type_of_move_call_input(called_packages, move_call, idx)?;
+                set_type(&mut literal_type, arg_type.to_owned())?;
+            }
+            (Command::TransferObjects(_), None) => {
+                set_type(&mut literal_type, Type::Address)?;
+            }
+
+            (Command::SplitCoins(_), Some(_)) => {
+                set_type(&mut literal_type, Type::U64)?;
+            }
+            (Command::MakeMoveVector(make_move_vector), Some(_)) => {
+                if let Some(ty) = &make_move_vector.type_ {
+                    let ty =
+                        sui_types::sui_sdk_types_conversions::type_tag_sdk_to_core(ty.clone())?;
+                    set_type(&mut literal_type, ty.into())?;
+                } else {
+                    return Err(RestError::new(
+                        axum::http::StatusCode::BAD_REQUEST,
+                        "unable to resolve literal as an unknown type",
+                    ));
+                }
+            }
+
+            // Invalid uses of Literal Arguments
+
+            // Pure arg can't be used as an object to transfer
+            (Command::TransferObjects(_), Some(_))
+            | (Command::Upgrade(_), _)
+            | (Command::MergeCoins(_), _)
+            | (Command::SplitCoins(_), None) => {
+                return Err(RestError::new(
+                    axum::http::StatusCode::BAD_REQUEST,
+                    "invalid use of literal",
+                ));
+            }
+
+            // bug in find_arg_uses
+            (Command::MakeMoveVector(_), None)
+            | (Command::Publish(_), _)
+            | (Command::MoveCall(_), None) => {
+                return Err(RestError::new(
+                    axum::http::StatusCode::INTERNAL_SERVER_ERROR,
+                    "error determining type of literal",
+                ));
+            }
+        }
+    }
+
+    literal_type.ok_or_else(|| {
+        RestError::new(
+            axum::http::StatusCode::BAD_REQUEST,
+            "unable to determine type of literal",
```

Comment:
can we provide the command / literal to offer more insight into what went wrong?

Reply:
Yes, there's a whole lot of polish that could be done on the error paths. I've sort of treated errors are throw away to get a well working happy path done but i'll spend some cycles on cleaning up errors